### PR TITLE
8244154: Update SunPKCS11 provider with PKCS11 v3.0 header files

### PR DESCRIPTION
**Pull request by valeriepeng**
Could someone please help review this PKCS#11 v3.0 header files update?

Changes are straight-forward as below:
1) Updated pkcs11.h, pkcs11f.h, pkcs11t.h to v3.0
2) Updated java side w/ the new constants definitions and name/error code mappings.

For the native headers, it's a direct copy of the official v3.0 headers except that I have to remove the tab space, and trailing white spaces due to JDK code requirement. I verified the result using 'diff -w'. As for the java side, the edit is based on the diff of native headers. I also commented out some of the unused native identifiers at java side.

I am adding the SHA-3 digests, signatures, and macs in a separate RFE and would need this one to be reviewed/integrated first.

Thanks,
Valerie

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8244154](https://bugs.openjdk.java.net/browse/JDK-8244154): Update SunPKCS11 provider with PKCS11 v3.0 header files

### Reviewers
 * [Weijun Wang](https://openjdk.java.net/census#weijun) (@wangweij - **Reviewer**) ⚠️ Review applies to 03f1b67af2aaee005e037ba54b70e58cee946c8f

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/917/head:pull/917`
`$ git checkout pull/917`

### CODE REVIEW

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java [hunk 1]
```diff
@@ -740,12 +783,14 @@ private static void addMGF(long id, String name) {
...
-        addMech(CKM_ACTI,                       "CKM_ACTI");
-        addMech(CKM_ACTI_KEY_GEN,               "CKM_ACTI_KEY_GEN");
+        addMech(CKM_SHA3_256_RSA_PKCS,          "CKM_SHA3_256_RSA_PKCS");
+        addMech(CKM_SHA3_384_RSA_PKCS,          "CKM_SHA3_384_RSA_PKCS");
+        addMech(CKM_SHA3_512_RSA_PKCS,          "CKM_SHA3_512_RSA_PKCS");
+        addMech(CKM_SHA3_256_RSA_PKCS_PSS,      "CKM_SHA3_256_RSA_PKCS_PSS");
+        addMech(CKM_SHA3_384_RSA_PKCS_PSS,      "CKM_SHA3_384_RSA_PKCS_PSS");
+        addMech(CKM_SHA3_512_RSA_PKCS_PSS,      "CKM_SHA3_512_RSA_PKCS_PSS");
+        addMech(CKM_SHA3_224_RSA_PKCS,          "CKM_SHA3_224_RSA_PKCS");
+        addMech(CKM_SHA3_224_RSA_PKCS_PSS,      "CKM_SHA3_224_RSA_PKCS_PSS");
```
Comment (haimaychao) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java:793]: It appears that you're arranging the addMech(with CKM_xxx) based on the mechanism values. How about the code from #773 to #793, move it up?

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java [hunk 2]
```diff
@@ -964,16 +1034,65 @@ private static void addMGF(long id, String name) {
...
+        addMech(CKM_X2RATCHET_DECRYPT,          "CKM_X2RATCHET_DECRYPT");
+        addMech(CKM_XEDDSA,                     "CKM_XEDDSA");
+        addMech(CKM_HKDF_DERIVE,                "CKM_HKDF_DERIVE");
+        addMech(CKM_HKDF_DATA,                  "CKM_HKDF_DATA");
+        addMech(CKM_HKDF_KEY_GEN,               "CKM_HKDF_KEY_GEN");
+        addMech(CKM_SALSA20_KEY_GEN,            "CKM_SALSA20_KEY_GEN");
+        addMech(CKM_SP800_108_COUNTER_KDF,      "CKM_SP800_108_COUNTER_KDF");
+        addMech(CKM_SP800_108_FEEDBACK_KDF,     "CKM_SP800_108_FEEDBACK_KDF");
+        addMech(CKM_SP800_108_DOUBLE_PIPELINE_KDF,
+                                         "CKM_SP800_108_DOUBLE_PIPELINE_KDF");
```
Comment (haimaychao) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java:1095]: same comment as above.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11Constants.java [hunk 3]
```diff
@@ -784,23 +911,80 @@
...
+    public static final long  CKM_XEDDSA                     = 0x00004029L;
+    public static final long  CKM_HKDF_DERIVE                = 0x0000402aL;
+    public static final long  CKM_HKDF_DATA                  = 0x0000402bL;
+    public static final long  CKM_HKDF_KEY_GEN               = 0x0000402cL;
+    public static final long  CKM_SALSA20_KEY_GEN            = 0x0000402dL;
+
+    public static final long  CKM_SP800_108_COUNTER_KDF      = 0x000003acL;
+    public static final long  CKM_SP800_108_FEEDBACK_KDF     = 0x000003adL;
+    public static final long  CKM_SP800_108_DOUBLE_PIPELINE_KDF = 0x000003aeL;
+
```
Comment (haimaychao) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11Constants.java:987]: Same comment.

**Review by haimaychao — COMMENTED** [review]
Changes look good. Only minor comments.

**Comment by wangweij** [comment]
Just curious, can the Java files be generated during the build process?

**Comment by wangweij** [comment]
The constants in PKCS11Exception are duplicated in PKCS11Constants.
```
            0x00000000,
```
vs
```
public static final long  CKR_OK                 = 0x00000000L;
```
Is there any way to simplify it?

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java (hunk 1 above)
Comment (valeriepeng) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java:793]: Yes, I will move it up and add a comment about overall ordering is based on PKCS11Constants class.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java (hunk 2 above)
Comment (valeriepeng) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java:1095]: Well, per the ordering in PKCS11Constants, these three lines are at the right place. Note that the ordering of CKM_ECDSA_SHA3_224 to CKM_EDDSA in pkcs11t.h is different from PKCS11Constants class, so I will add the comment about the general ordering following PKCS11Constants class and keep them here.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11Constants.java (hunk 3 above)
Comment (valeriepeng) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11Constants.java:987]: These three are just by themselves, so unless you feel strongly about, I prefer just leave them here which matches the ordering of pkcs11t.h, i.e. right before the CKM_VENDOR_DEFINED line.

**Comment by valeriepeng** [comment]
> 
> 
> Just curious, can the Java files be generated during the build process?

Hmm, maybe, by the java files, do you just mean PKCS11Constants class or more? I am not familiar with how to generate Java files during the build process, can you share some pointers? I can look into them for possible future enhancement.

**Comment by valeriepeng** [comment]
> 
> 
> The constants in PKCS11Exception are duplicated in PKCS11Constants.
> 
> ```
>             0x00000000,
> ```
> 
> vs
> 
> ```
> public static final long  CKR_OK                 = 0x00000000L;
> ```
> 
> Is there any way to simplify it?

One defines the value, the other defines the displayed String. I agree that the way it's currently done is a bit outdated and may be error prone. We can re-visit this after finishing other pending PKCS11 RFEs..

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java (hunk 2 above)
Comment (haimaychao) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/Functions.java:1095]: Sounds good that a comment will be added.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11Constants.java (hunk 3 above)
Comment (haimaychao) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11Constants.java:987]: Just thought they could be moved like CKM_ECDSA_SHA3_224 to CKM_EDDSA (which are not matching to the ordering of pkcs11t.h), so we keep the ordering in PKCS11Constants. It's fine if you'd prefer as is.

**Comment by wangweij** [comment]
> > The constants in PKCS11Exception are duplicated in PKCS11Constants.
> > ```
> >             0x00000000,
> > ```
> > 
> > 
> > vs
> > ```
> > public static final long  CKR_OK                 = 0x00000000L;
> > ```
> > 
> > 
> > Is there any way to simplify it?
> 
> One defines the value, the other defines the displayed String. I agree that the way it's currently done is a bit outdated and may be error prone. We can re-visit this after finishing other pending PKCS11 RFEs..

Since the value is already defined PKCS11Constants, we can use it in PKCSException and does not need to write 0x00000000 anymore.

**Comment by valeriepeng** [comment]
> 
> 
> > > The constants in PKCS11Exception are duplicated in PKCS11Constants.
> > > ```
> > >             0x00000000,
> > > ```
> > > 
> > > 
> > > vs
> > > ```
> > > public static final long  CKR_OK                 = 0x00000000L;
> > > ```
> > > 
> > > 
> > > Is there any way to simplify it?
> > 
> > 
> > One defines the value, the other defines the displayed String. I agree that the way it's currently done is a bit outdated and may be error prone. We can re-visit this after finishing other pending PKCS11 RFEs..
> 
> Since the value is already defined PKCS11Constants, we can use it in PKCSException and does not need to write 0x00000000 anymore.

Ok, I see what you are saying now. Will change it to use the PKCS11Constants constants.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11Constants.java (hunk 3 above)
Comment (valeriepeng) [reply src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11Constants.java:987]: Thanks, I have slight preference for as is.

**Comment by wangweij** [comment]
> > Just curious, can the Java files be generated during the build process?
> 
> Hmm, maybe, by the java files, do you just mean PKCS11Constants class or more? I am not familiar with how to generate Java files during the build process, can you share some pointers? I can look into them for possible future enhancement.

Well, any file that can was a rewrite of its C sibling.

SunEC's finite field arithmetic code is generated in the build process. The generation tool is https://github.com/openjdk/jdk/blob/fd28aad72d3a13fbc2eb13293d40564e471414f3/make/jdk/src/classes/build/tools/intpoly/FieldGen.java, and it's called by https://github.com/openjdk/jdk/blob/1cb7df63e7015da568b67c55a7a4468500ea564d/make/modules/java.base/Gensrc.gmk#L113 and the files are generated in the `$BUILD_OUTPUT/support/gensrc` directory. These files will be automatically merged with the human-written sources files in `src` of the same package at build time.

**Comment by wangweij** [comment]
I cannot comment on the unchanged lines in PKCS11Constants.java, but the class spec might be enhanced a little.

1. CK_SESSION_HANDLE appears twice.
2. The following appears in pkcs11t.h and I wonder if they can also be added here:
```
typedef CK_ULONG CK_OTP_PARAM_TYPE;
typedef CK_OTP_PARAM_TYPE CK_PARAM_TYPE; /* backward compatibility */
typedef CK_ULONG CK_GENERATOR_FUNCTION;
typedef CK_ULONG CK_JAVA_MIDP_SECURITY_DOMAIN;
typedef CK_ULONG CK_CERTIFICATE_CATEGORY;
typedef CK_ULONG CK_PROFILE_ID;
typedef CK_ULONG CK_PRF_DATA_TYPE;
typedef CK_MECHANISM_TYPE CK_SP800_108_PRF_TYPE;
typedef CK_ULONG CK_SP800_108_DKM_LENGTH_METHOD;
typedef CK_ULONG CK_X3DH_KDF_TYPE;
typedef CK_ULONG CK_X2RATCHET_KDF_TYPE;
typedef CK_ULONG CK_XEDDSA_HASH_TYPE;
```
I also found 2 bugs in pkcs11t.h. `CK_GCM_MESSAGE_PARAMS_PTR` and `CK_CCM_MESSAGE_PARAMS_PTR` are not defined as `CK_PTR` of the corresponding data type. Maybe you can report this to upstream?

**Comment by valeriepeng** [comment]
> 
> 
> https://github.com/openjdk/jdk/blob/0b37b821a10325d9083c23130e4b8921812ed9c5/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/wrapper/PKCS11Constants.java#L54
> 
> I cannot add comments to unchanged lines in PKCS11Constants.java (there's no + sign on the line numbers), but the class-level comment (starting from line 56) can also be enhanced a little.

Sure, I will update here as well.
 
>     1. CK_SESSION_HANDLE appears twice.
> 
>     2. The following appears in pkcs11t.h and I wonder if they can also be added here:
> 
> 
> ```
> typedef CK_ULONG CK_OTP_PARAM_TYPE;
> typedef CK_OTP_PARAM_TYPE CK_PARAM_TYPE; /* backward compatibility */
> typedef CK_ULONG CK_GENERATOR_FUNCTION;
> typedef CK_ULONG CK_JAVA_MIDP_SECURITY_DOMAIN;
> typedef CK_ULONG CK_CERTIFICATE_CATEGORY;
> typedef CK_ULONG CK_PROFILE_ID;
> typedef CK_ULONG CK_PRF_DATA_TYPE;
> typedef CK_MECHANISM_TYPE CK_SP800_108_PRF_TYPE;
> typedef CK_ULONG CK_SP800_108_DKM_LENGTH_METHOD;
> typedef CK_ULONG CK_X3DH_KDF_TYPE;
> typedef CK_ULONG CK_X2RATCHET_KDF_TYPE;
> typedef CK_ULONG CK_XEDDSA_HASH_TYPE;
> ```
> 
> I also found 2 bugs in pkcs11t.h. `CK_GCM_MESSAGE_PARAMS_PTR` and `CK_CCM_MESSAGE_PARAMS_PTR` are not defined as `CK_PTR` of their corresponding data types. Maybe you can report this to upstream?

Right, these two looks wrongly defined. I will send a comment about this to the Oasic PKCS11 TC.
Thanks~

**Review by wangweij — APPROVED** [review]

**Comment by haimaychao** [comment]
Incr webrev looks good.

**Pull request closed without merging**